x-pack/plugin/mapper-aggregate-metric/src/main/java/org/elasticsearch/xpack/aggregatemetric/mapper/AggregateDoubleMetricFieldMapper.java

```diff
@@ -0,0 +1,481 @@
+        @Override
+        public Object valueForDisplay(Object value) {
+            return delegateFieldType().valueForDisplay(value);
+        }
+
+        @Override
+        public DocValueFormat docValueFormat(String format, ZoneId timeZone) {
+            return delegateFieldType().docValueFormat(format, timeZone);
+        }
+
+        @Override
+        public Relation isFieldWithinQuery(
+            IndexReader reader,
+            Object from, Object to,
+            boolean includeLower, boolean includeUpper,
+            ZoneId timeZone, DateMathParser dateMathParser, QueryRewriteContext context) throws IOException {
+            return delegateFieldType().isFieldWithinQuery(reader, from, to, includeLower, includeUpper, timeZone, dateMathParser, context);
+        }
+
+        @Override
+        public IndexFieldData.Builder fielddataBuilder(String fullyQualifiedIndexName) {
+            return delegateFieldType().fielddataBuilder(fullyQualifiedIndexName);
+        }
+
+    }
+
+    private final EnumMap<Metric, NumberFieldMapper> metricFieldMappers;
+
+    private Explicit<Boolean> ignoreMalformed;
+
+    /** A set of metrics supported */
+    private Explicit<Set<Metric>> metrics;
+
+    /** The default metric to be when querying this field type */
+    protected Explicit<Metric> defaultMetric;
+
+    private AggregateDoubleMetricFieldMapper(String simpleName, MappedFieldType fieldType, MappedFieldType defaultFieldType,
+                                             Settings indexSettings, MultiFields multiFields, Explicit<Boolean> ignoreMalformed,
+                                             Explicit<Set<Metric>> metrics, Explicit<Metric> defaultMetric,
+                                             CopyTo copyTo, EnumMap<Metric, NumberFieldMapper> metricFieldMappers) {
+        super(simpleName, fieldType, defaultFieldType, indexSettings, multiFields, copyTo);
+        this.ignoreMalformed = ignoreMalformed;
+        this.metrics = metrics;
+        this.defaultMetric = defaultMetric;
+        this.metricFieldMappers = metricFieldMappers;
+    }
+
+    @Override
+    public AggregateDoubleMetricFieldType fieldType() {
+        return (AggregateDoubleMetricFieldType) super.fieldType();
+    }
+
+    @Override
+    protected String contentType() {
+        return fieldType.typeName();
+    }
+
+    @Override
+    protected AggregateDoubleMetricFieldMapper clone() {
+        return (AggregateDoubleMetricFieldMapper) super.clone();
+    }
+
+    @Override
+    public Iterator<Mapper> iterator() {
+        List<Mapper> mappers = new ArrayList<>(metricFieldMappers.values());
+        return mappers.iterator();
+    }
+
+    @Override
+    protected void parseCreateField(ParseContext context, List<IndexableField> fields) throws IOException {
+        if (context.externalValueSet() == true) {
+            throw new IllegalArgumentException("Field [" + name() + "] of type [" + typeName() + "] can't be used in multi-fields");
+        }
+
```

Comment:
should we handle the case when the value is an array of metrics?

Reply:
I cannot think of a case that an array aggregate metrics can be useful. Therefore, we don't plan to support it

Reply:
It is ok not to support it but we should capture that case and throw a meaningful error?

Reply:
See:   https://github.com/elastic/elasticsearch/blob/eade4f03f426447fea6148f7cf482ee0baee4cd3/x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/mapper/HistogramFieldMapper.java#L413